File: app/src/main/java/com/example/drawing/PaintView.java
package com.example.drawing;

import android.animation.Animator;
import android.animation.ObjectAnimator;
import android.animation.ValueAnimator;
import android.content.Context;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.DashPathEffect;
import android.graphics.Paint;
import android.graphics.Path;
import android.graphics.PathEffect;
import android.graphics.PathMeasure;
import android.util.AttributeSet;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;

import java.util.ArrayList;

public class PaintView extends View {
    public ViewGroup.LayoutParams layoutParams;

    float width = 1920;
    float height = 1080;
    public PaintView(Context context) {
        super(context);

        layoutParams = new ViewGroup.LayoutParams((int) width, (int) height);
    }

    public PaintView(Context context, AttributeSet attrs) {
        super(context, attrs);
    }

    public PaintView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
    }

    private Paint brushSample = new Paint();
    private Paint brushStroke = new Paint();
    private Paint brushDraw = new Paint();

    private Path pathSample = new Path();
    private Path pathStroke = new Path();
    private Path pathDraw = new Path();
    public void initBrush() {
        brushSample.setAntiAlias(true);
        brushSample.setStyle(Paint.Style.STROKE);
        brushSample.setStrokeJoin(Paint.Join.ROUND);
        brushSample.setStrokeCap(Paint.Cap.ROUND);
        brushSample.setColor(Color.CYAN);
        brushSample.setStrokeWidth(30f);

        brushStroke.setAntiAlias(true);
        brushStroke.setStyle(Paint.Style.STROKE);
        brushStroke.setStrokeJoin(Paint.Join.ROUND);
        brushStroke.setStrokeCap(Paint.Cap.ROUND);
        brushStroke.setColor(Color.BLACK);
        brushStroke.setStrokeWidth(45f);

        brushDraw.setAntiAlias(true);
        brushDraw.setStyle(Paint.Style.STROKE);
        brushDraw.setStrokeJoin(Paint.Join.ROUND);
        brushDraw.setStrokeCap(Paint.Cap.ROUND);
        brushDraw.setColor(Color.YELLOW);
        brushDraw.setStrokeWidth(30f);

        pathSample.setFillType(Path.FillType.EVEN_ODD);
        pathStroke.setFillType(Path.FillType.EVEN_ODD);
        pathDraw.setFillType(Path.FillType.EVEN_ODD);
    }


    boolean isDrawSample = false;
    public void drawSample(ArrayList<PathModel> listPath) {
        for (int i = 0; i < listPath.size(); i++) {
            pathSample.moveTo(listPath.get(i).getPath().get(0).getX(), listPath.get(i).getPath().get(0).getY());
            pathStroke.moveTo(listPath.get(i).getPath().get(0).getX(), listPath.get(i).getPath().get(0).getY());
            for (int j = 1; j < listPath.get(i).getPath().size(); j++) {
                pathSample.lineTo(listPath.get(i).getPath().get(j).getX(), listPath.get(i).getPath().get(j).getY());
                pathStroke.lineTo(listPath.get(i).getPath().get(j).getX(), listPath.get(i).getPath().get(j).getY());
            }
        }
        isDrawSample = true;
        invalidate();
    }

    boolean isDrawAnimation = false;
    public void drawAnimation(ArrayList<PathModel> listPath) {
        for (int i = 0; i < listPath.size(); i++) {
            pathDraw.moveTo(listPath.get(i).getPath().get(0).getX(), listPath.get(i).getPath().get(0).getY());
            for (int j = 1; j < listPath.get(i).getPath().size(); j++) {
                pathDraw.lineTo(listPath.get(i).getPath().get(j).getX(), listPath.get(i).getPath().get(j).getY());
            }
        }

        PathMeasure measure = new PathMeasure(pathDraw, true);
        float length = measure.getLength();

        ObjectAnimator animator = ObjectAnimator.ofFloat(this, "phase", 1.0f, 0.0f);
        animator.setDuration(5000);
        animator.addUpdateListener(valueAnimator -> {
            brushDraw.setPathEffect(createPathEffect(length, (Float) valueAnimator.getAnimatedValue(), 0.0f));
            invalidate();//will calll onDraw
        });
        animator.addListener(new Animator.AnimatorListener() {
            @Override
            public void onAnimationStart(Animator animator) {
                isDrawAnimation = true;
            }

            @Override
            public void onAnimationEnd(Animator animator) {
                
            }

            @Override
            public void onAnimationCancel(Animator animator) {

            }

            @Override
            public void onAnimationRepeat(Animator animator) {

            }
        });
        animator.start();
    }

    public void setPhase(float phase) {
        Log.d("qqDebug","setPhase called with:" + phase);
    }

    private static PathEffect createPathEffect(float pathLength, float phase, float offset) {
        return new DashPathEffect(new float[] { pathLength, pathLength },
                Math.max(phase * pathLength, offset));
    }

    @Override
    protected void onDraw(Canvas canvas) {
        canvas.translate(0,getHeight());
        canvas.scale(1f, -1f);
        super.onDraw(canvas);

        if (isDrawSample) {
            canvas.drawPath(pathStroke, brushStroke);
            canvas.drawPath(pathSample, brushSample);
            isDrawSample = false;
        }

        if (isDrawAnimation) {
            canvas.drawPath(pathDraw, brushDraw);
        }
    }
}

File: app/src/main/java/com/example/drawing/PaintView.kt
package com.example.drawing

import android.animation.Animator
import android.animation.ObjectAnimator
import android.animation.ValueAnimator
import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.util.Log
import android.view.View
import com.google.gson.GsonBuilder
import org.json.JSONArray
import java.util.ArrayList

class PaintView : View {
    constructor(context: Context) : super(context)

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int) : super(
        context,
        attrs,
        defStyleAttr
    )

    private val brushSample = Paint()
    private val brushStroke = Paint()
    private val brushDraw = ArrayList<Paint>()
    private val pathSample = Path()
    private val pathStroke = Path()
    private val pathDraw = ArrayList<Path>()
    fun initBrush() {
        brushSample.isAntiAlias = true
        brushSample.style = Paint.Style.STROKE
        brushSample.strokeJoin = Paint.Join.ROUND
        brushSample.strokeCap = Paint.Cap.ROUND
        brushSample.color = Color.CYAN
        brushSample.strokeWidth = 30f
        brushStroke.isAntiAlias = true
        brushStroke.style = Paint.Style.STROKE
        brushStroke.strokeJoin = Paint.Join.ROUND
        brushStroke.strokeCap = Paint.Cap.ROUND
        brushStroke.color = Color.BLACK
        brushStroke.strokeWidth = 45f
        pathSample.fillType = Path.FillType.EVEN_ODD
        pathStroke.fillType = Path.FillType.EVEN_ODD
    }

    var isDrawSample = false

    fun drawSample(jsArray: JSONArray) {
        val listPathDraw : ArrayList<PathModel> = arrayListOf()
        for (i in 0 until jsArray.length()) {
            val jsonObject = jsArray.getJSONObject(i)
            val gson = GsonBuilder().create()
            val model : ArrayList<CoordinateModel> = arrayListOf()
            gson.fromJson(jsonObject.getString("lsVectorsInStroke"),Array<CoordinateModel>::class.java).forEach {
                model.add(CoordinateModel(it.x , it.y, it.z))
            }

            listPathDraw.add(PathModel(model))
        }

        measurePathScale(listPathDraw)
        for (i in listPathDraw.indices) {
            pathSample.moveTo(listPathDraw[i].path[0].x, listPathDraw[i].path[0].y)
            pathStroke.moveTo(listPathDraw[i].path[0].x, listPathDraw[i].path[0].y)
            for (j in 1 until listPathDraw[i].path.size) {
                pathSample.lineTo(listPathDraw[i].path[j].x, listPathDraw[i].path[j].y)
                pathStroke.lineTo(listPathDraw[i].path[j].x, listPathDraw[i].path[j].y)
            }
        }
        isDrawSample = true
        invalidate()
    }

    var isDrawAnimation = false
    var isAnimationFinish = 0
    var currentPath = 0
    var length = 0f
    fun drawAnimation(jsArray: JSONArray) {
        val listPathDraw : ArrayList<PathModel> = arrayListOf()
        for (i in 0 until jsArray.length()) {
            val jsonObject = jsArray.getJSONObject(i)
            val gson = GsonBuilder().create()
            val model : ArrayList<CoordinateModel> = arrayListOf()
            gson.fromJson(jsonObject.getString("lsVectorsInStroke"),Array<CoordinateModel>::class.java).forEach {
                model.add(CoordinateModel(it.x , it.y, it.z))
            }

            listPathDraw.add(PathModel(model))
        }

        if (currentPath >= listPathDraw.size) return

        for (i in listPathDraw.indices) {
            val subPath = Path()
            subPath.fillType = Path.FillType.EVEN_ODD
            subPath.moveTo(listPathDraw[i].path[0].x, listPathDraw[i].path[0].y)
            for (j in 1 until listPathDraw[i].path.size) {
                subPath.lineTo(listPathDraw[i].path[j].x, listPathDraw[i].path[j].y)
            }
            pathDraw.add(subPath)
            val brush = Paint()
            brush.isAntiAlias = true
            brush.style = Paint.Style.STROKE
            brush.strokeJoin = Paint.Join.ROUND
            brush.strokeCap = Paint.Cap.ROUND
            brush.color = Color.YELLOW
            brush.strokeWidth = 30f
            brushDraw.add(brush)
        }
        val measure = PathMeasure(pathDraw[currentPath], false)
        length = measure.length
        val animator = ObjectAnimator.ofFloat(this, "phase", 1.0f, 0.0f)
        animator.duration = 5000
        animator.addUpdateListener { valueAnimator: ValueAnimator ->
            brushDraw[currentPath].pathEffect =
                createPathEffect(length, valueAnimator.animatedValue as Float, 0.0f)
            invalidate() //will calll onDraw
        }
        animator.addListener(object : Animator.AnimatorListener {
            override fun onAnimationStart(animator: Animator) {
                isDrawAnimation = true
            }

            override fun onAnimationEnd(animator: Animator) {
                isDrawAnimation = false
                if (currentPath < pathDraw.size - 1) {
                    currentPath++
                    val measure = PathMeasure(pathDraw[currentPath], false)
                    length = measure.length
                    Log.d("qqDebug", "pathLength: $length")
                    isAnimationFinish = currentPath
                    drawAnimation(jsArray)
                }
            }

            override fun onAnimationCancel(animator: Animator) {}
            override fun onAnimationRepeat(animator: Animator) {}
        })
        animator.start()
    }

    fun setPhase(phase: Float) {
//        Log.d("qqDebug","setPhase called with:" + phase);
    }

    var scaleRatio = 1f
    var screenW = 0
    var screenH = 0
    var offsetRatio = 0f
    fun setScreenMeasure(screenW: Int, screenH: Int) {
        this.screenW = screenW
        this.screenH = screenH
    }

    var pathWidth = 1f
    var pathHeight = 1f
    var maxX = 1f
    var maxY = 1f
    private fun measurePathScale(listPath: ArrayList<PathModel>) {
        val xList : ArrayList<Float> = arrayListOf()
        val yList : ArrayList<Float> = arrayListOf()

        for (i in 0 until listPath.size) {
            xList.add(listPath[i].path.minOf { axes -> axes.x })
            xList.add(listPath[i].path.maxOf { axes -> axes.x })

            yList.add(listPath[i].path.minOf { axes -> axes.y })
            yList.add(listPath[i].path.maxOf { axes -> axes.y })
        }

        maxX = xList.maxOf { float -> float }
        maxY = yList.maxOf { float -> float }
        pathWidth = xList.maxOf { float -> float } - xList.minOf { float -> float }
        pathHeight = yList.maxOf { float -> float } - yList.minOf { float -> float }

        Log.d("qqDebug", "pathW: $pathWidth\npathH: $pathHeight")
    }


    fun initRatio(viewW: Int, viewH: Int) {
        Log.d("qqDebug", "maxX: $maxX\nmaxY: $maxY")
        if (screenW > screenH) {
            scaleRatio = viewW.toFloat() / pathWidth - /*for padding*/(viewW.toFloat() / pathWidth) / 9f
            offsetRatio = scaleRatio * -95f
//            offsetRatio = scaleRatio * (screenW - maxX)
        } else {
            scaleRatio = viewH.toFloat() / pathHeight - /*for padding*/(viewW.toFloat() / pathHeight) / 9f
            offsetRatio = scaleRatio * -95f
//            offsetRatio = scaleRatio * (screenH - maxY)
        }

        Log.d("qqDebug", "sW: $screenW, sH: $screenH")
        Log.d("qqDebug", "vW: $viewW, vH: $viewH")
        Log.d("qqDebug", "scaleRatio: $scaleRatio")
        Log.d("qqDebug", "offsetRatio: $offsetRatio")
    }

    override fun onDraw(canvas: Canvas) {
        canvas.translate(0f, height.toFloat())
        canvas.scale(1f, -1f)
        initRatio(width, height)
        canvas.translate(offsetRatio * 4, offsetRatio * 0.9f)
        canvas.scale(scaleRatio, scaleRatio)
        super.onDraw(canvas)
        canvas.drawPath(pathStroke, brushStroke)
        canvas.drawPath(pathSample, brushSample)
        Log.d("qqDebug", "draw times: $isAnimationFinish")
        when (isAnimationFinish) {
            1 -> canvas.drawPath(pathDraw[0], brushDraw[0])
            2 -> {
                canvas.drawPath(pathDraw[0], brushDraw[0])
                canvas.drawPath(pathDraw[1], brushDraw[1])
            }
            3 -> {
                canvas.drawPath(pathDraw[0], brushDraw[0])
                canvas.drawPath(pathDraw[1], brushDraw[1])
                canvas.drawPath(pathDraw[2], brushDraw[2])
            }
            else -> {
            }
        }
        if (isDrawAnimation) {
            canvas.drawPath(pathDraw[currentPath], brushDraw[currentPath])
        }
        if (!pathDraw.isEmpty()) {
            canvas.drawPath(pathDraw[currentPath], brushDraw[currentPath])
        }
    }

    companion object {
        private fun createPathEffect(pathLength: Float, phase: Float, offset: Float): PathEffect {
            return DashPathEffect(
                floatArrayOf(pathLength, pathLength),
                Math.max(phase * pathLength, offset)
            )
        }
    }
}